grant management initial commit

guymoyo opened the pull request:
In the scope of this PR, I would like to show the conception of Grant Management.
Please note, it's not totally complete. If the conception is correct (I will go on then), if there are some advice or propositions - feel free to comment. Thank you!

Draft design: [https://github.com/keycloak/keycloak-community/pull/265](https://github.com/keycloak/keycloak-community/pull/265)

spec: [https://openid.net/specs/fapi-grant-management-01.html](https://openid.net/specs/fapi-grant-management-01.html)

tnorimat commented:
Hello @guymoyo  ,
Thank you for your implementation. I will review it and add comments afterwards.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java:
```diff
@@ -245,6 +246,13 @@ public Object issueUserInfo() {
         return endpoint;
     }
 
+    @Path("grants")
```
As for the way of implementing Endpoint class, It might be preferable to follow `DeviceEndpoint` and `DeviceEndpointFactory` rather than implementing it in this service class.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java:
```diff
@@ -160,6 +160,9 @@ public Object getConfig() {
         config.setBackchannelLogoutSupported(true);
         config.setBackchannelLogoutSessionSupported(true);
 
+        config.setGrantIdSupported(true);
```
According to https://openid.net/specs/fapi-grant-management-01.html#section-5.1 , it might be "none", "optional" or "always".
As mentioned earlier, "optional" might be appropriate.

tnorimat commented on core/src/main/java/org/keycloak/protocol/oidc/representations/OIDCConfigurationRepresentation.java:
```diff
@@ -142,8 +142,30 @@
     private Boolean backchannelLogoutSessionSupported;
 
+    @JsonProperty("grant_id_supported")
+    private Boolean grantIdSupported;
```
According to https://openid.net/specs/fapi-grant-management-01.html#section-5.1 , it might be "none", "optional" or "always".
As mentioned later, "optional" might be appropriate.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthorizationEndpointRequest.java:
```diff
@@ -50,6 +50,10 @@
 
+    String grantId;
+
+    public String getGrantId() { return grantId; }
```
By keycloak's convension, it might be better,
```
public String getGrantId() {
    return grantId;
}
```

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/grants/management/endpoints/GrantEndpoint.java:
```diff
@@ -0,0 +1,321 @@
+
+
+
+    public GrantEndpoint(RealmModel realm, EventBuilder event, org.keycloak.protocol.oidc.TokenManager tokenManager) {
```
IMO, it might be better to authenticate the client who send grant management request (query, revoke) with an access token.
Moreover, it might better to confirm that the authenticated client is the one to whom keycloak issued the access token.

The current endpoint allows only simple bearer token usage so that anyone who steals other    legitimate client's access token can revoke their grants.

guymoyo replied:
agree

tnorimat commented on services/src/main/java/org/keycloak/services/resources/LoginActionsService.java:
```diff
@@ -49,16 +51,7 @@
-import org.keycloak.models.RealmModel;
-import org.keycloak.models.UserConsentModel;
-import org.keycloak.models.UserModel;
+import org.keycloak.models.*;
```
By keycloak's convention, it might be better not use wildcard "*" in import directive.

tnorimat commented on model/jpa/src/main/java/org/keycloak/grant/jpa/impl/GrantServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+
+        em.remove(userGrant);
+        em.flush();
+        return true;
```
If a grant is revoked, is an access token and refresh token that have been issued when this grant has been created still active or not?
If not, a client could still use these tokens even if its related grant was revoked.

It might be better to clarify the relationship among
existing Consent, existing UserSession, and newly introduced Grant. 

If a grant is created, modified or revoked, what happens to a consent and a user session.
If a consent is created, modified or revoked, what happens to a grant.
If a user session is created, modified or revoked, what happens to a grant.

guymoyo replied:
Normally when a grant is revoked, an access token and refresh token that has been issued when this grant has been created should no longer be active. Still looking at how to implement that.

IMO  If we delete/modify a grant, there is no impact on consent neither userSession.
if we delete/modify userSession, there is no impact on a grant
if we delete a consent, all grant with the same client and user should be deleted

WDYT?

tnorimat replied:
@guymoyo Generally, I agree with you.

tnorimat commented on model/jpa/src/main/java/org/keycloak/grant/jpa/UserGrantEntity.java:
```diff
@@ -0,0 +1,196 @@
+@Entity
+@Table(name="USER_GRANT")
+@IdClass(UserGrantEntity.Key.class)
+public class UserGrantEntity {
```
By recent keycloak's convention, it might be better not to add or modify JPA entity because it causes a schema change. For example, Client Policies/Profiles avoid adding entities by storing JSON string itself to realm's attribute (existing key-value store).

guymoyo replied:
I see, we can stored the grants as the realm's attribute with the key "grants-extension". And the grants can be represented as follows

{
	"grants": [
	
		{
			"grantId" : ""
			"clientId" : ""
			"userId" : ""
			"authorization_details" : {}
			"scopes" : []
			"claims" : []
			"createdDate" ""
			"lastUpdatedDate" ""
		},

	]
}

WDYT?

guymoyo replied:
it's strange, realm's attribute can not get a long value:

`Caused by: org.h2.jdbc.JdbcSQLException: Value too long for column "VALUE NVARCHAR(255)": "'{""grants"":[{""grantId"":""lODVWR2zz2AfWW2CeukYdFLjvLUjKNtC41U8EYCDPsQ"",""clientId"":""client-1"",""userId"":""d1419007-7f92-43a8-9c82-4d7... (268)"; SQL statement:
update REALM_ATTRIBUTE set VALUE=? where NAME=? and REALM_ID=? [22001-197]`

tnorimat replied:
@guymoyo Could you rebase your branch on the latest keycloak master branch? It included extending Realm Attribute's size. You can found it in detail on the following :
https://github.com/keycloak/keycloak/pull/7780/files#diff-3374e8fe71e9cf4c16ff00d889e8305734333b2c059d8673aaa1060750d0ac79

guymoyo replied:
Thanks, I looked at this PR. It's clear however if I rebase on the latest keycloak master branch, should i not have a lot of conflicts / commits when trying to merge on tnorimat / keycloak master branch because it is not yet up to date?

tnorimat replied:
@guymoyo How about the following?

1. I will rebase tnorimat/master on keycloak/master
2. you will rebase your branch on keycloak/master
3. you will push -f your rebased branch to tnorimat

But I'm not sure it works well...

tnorimat commented on model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserConsentEntity.java:
```diff
@@ -17,6 +17,8 @@
 package org.keycloak.models.jpa.entities;
 
+import org.keycloak.grant.jpa.UserGrantEntity;
+
```
By recent keycloak's convention, it might be better not to add or modify JPA entity because it causes a schema change. For example, Client Policies/Profiles avoid adding entities by storing JSON string itself to realm's attribute (existing key-value store).

guymoyo replied:
will be remove and stored as described above

tnorimat commented on model/jpa/src/main/resources/META-INF/jpa-changelog-14.0.0.xml:
```diff
@@ -0,0 +1,46 @@
+
+    <changeSet author="keycloak" id="add-grant-ext">
+
+        <createTable tableName="USER_GRANT">
```
By recent keycloak's convention, it might be better not to add or modify JPA entity because it causes a schema change. For example, Client Policies/Profiles avoid adding entities by storing JSON string itself to realm's attribute (existing key-value store).

guymoyo replied:
will be remove and stored as described above

tnorimat commented on model/jpa/src/main/resources/META-INF/jpa-changelog-master.xml:
```diff
@@ -69,5 +69,6 @@
     <include file="META-INF/jpa-changelog-11.0.0.xml"/>
     <include file="META-INF/jpa-changelog-12.0.0.xml"/>
     <include file="META-INF/jpa-changelog-13.0.0.xml"/>
+    <include file="META-INF/jpa-changelog-14.0.0.xml"/>
```
By recent keycloak's convention, it might be better not to add or modify JPA entity because it causes a schema change. For example, Client Policies/Profiles avoid adding entities by storing JSON string itself to realm's attribute (existing key-value store).

guymoyo replied:
will be remove and stored as described above

tnorimat commented on model/jpa/src/main/resources/META-INF/persistence.xml:
```diff
@@ -43,6 +43,7 @@
         <class>org.keycloak.models.jpa.entities.IdentityProviderMapperEntity</class>
         <class>org.keycloak.models.jpa.entities.ProtocolMapperEntity</class>
         <class>org.keycloak.models.jpa.entities.UserConsentEntity</class>
+        <class>org.keycloak.grant.jpa.UserGrantEntity</class>
```
By recent keycloak's convention, it might be better not to add or modify JPA entity because it causes a schema change. For example, Client Policies/Profiles avoid adding entities by storing JSON string itself to realm's attribute (existing key-value store).

guymoyo replied:
will be remove and stored as described above

tnorimat commented on model/map/src/main/java/org/keycloak/models/map/user/UserConsentEntity.java:
```diff
@@ -37,6 +37,7 @@
     private Long createdDate;
     private Long lastUpdatedDate;
     private boolean updated;
+    private String userConsentId;
```
By recent keycloak's convention, it might be better not to add or modify JPA entity because it causes a schema change. For example, Client Policies/Profiles avoid adding entities by storing JSON string itself to realm's attribute (existing key-value store).

guymoyo replied:
will be remove and stored as described above

tnorimat commented on server-spi-private/src/main/java/org/keycloak/models/GrantSpi.java:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
```
might be 2021.

tnorimat commented on server-spi/src/main/java/org/keycloak/models/UserGrantModel.java:
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
```
might be 2021.

tnorimat commented on server-spi-private/src/main/java/org/keycloak/models/GrantServiceProviderFactory.java:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
```
might be 2021.

tnorimat commented on server-spi-private/src/main/java/org/keycloak/models/GrantService.java:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
```
might be 2021.

tnorimat commented on model/jpa/src/main/resources/META-INF/services/org.keycloak.models.GrantServiceProviderFactory:
```diff
@@ -0,0 +1,18 @@
+#
+# Copyright 2019 Red Hat, Inc. and/or its affiliates
```
might be 2021.

tnorimat commented on model/jpa/src/main/java/org/keycloak/grant/jpa/impl/GrantServiceProviderFactoryImpl.java:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
```
might be 2021.

tnorimat commented on model/jpa/src/main/java/org/keycloak/grant/jpa/impl/GrantServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
```
might be 2021.

tnorimat commented on model/jpa/src/main/java/org/keycloak/grant/jpa/UserGrantEntity.java:
```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates
```
might be 2021.

tnorimat reviewed (commented):
@guymoyo I've reviewed this PR and add some review comments. Could you check them?

guymoyo commented on core/src/main/java/org/keycloak/representations/AccessTokenResponse.java:
```diff
@@ -61,6 +61,18 @@
     @JsonProperty("scope")
     protected String scope;
 
+    // Grant Management for OAuth 2.0  : grant_id MUST be returned in the response from Token Endpoint
```
true

guymoyo commented on core/src/main/java/org/keycloak/protocol/oidc/representations/OIDCConfigurationRepresentation.java:
```diff
@@ -142,8 +142,30 @@
+    }
+
+    public void setGrantIdSupported(Boolean grantIdSupported) {
+        this.grantIdSupported = grantIdSupported;
```
yes, correct.

tnorimat commented on model/jpa/src/main/java/org/keycloak/grant/jpa/impl/GrantServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+
+
+    @Override
+    public boolean revokeGrantByGrantId(RealmModel realm, String grandId, String userId) {
```
@guymoyo AFAIK, there is no trivial solution. If you decide such that a newly introduced Grant corresponds with an Authenticated Client Session , you can take the same way as in a refresh token's token revocation. Please refer to `TokenRevocationEndpoint`. 
When receiving a refresh token on this endpoint, keycloak revokes corresponding Authenticated Client Session which leads to revoking a refresh token.

The way depends on the determination of the relationships between Grant, User Session, Authenticated Client Session, Access Token and Refresh Token. The simple way is to define such that 1 Grant corresponds with 1 Authenticated Client Session.

Keycloak does not manage access tokens and refresh tokens one by one. Instead of that, keycloak manages User Session, Authenticated Client Session and its latest updated refresh token. keycloak manages a consent per (user, client) pair.

guymoyo replied:
I see, a Grant with a corresponding Authenticated Client Session, we can revoke all the Refresh Token. But the Access Token will still be valid (although for a defined limited frame of time).

The only way I see now for revoking Access Token is to keep the AccessTokenId along with the grant.

guymoyo commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/UserCacheSession.java:
```diff
@@ -729,6 +729,7 @@ private UserConsentModel toConsentModel(RealmModel realm, CachedUserConsent cach
         UserConsentModel consentModel = new UserConsentModel(client);
         consentModel.setCreatedDate(cachedConsent.getCreatedDate());
         consentModel.setLastUpdatedDate(cachedConsent.getLastUpdatedDate());
+        consentModel.setUserConsentId(consentModel.getUserConsentId());
```
true, I think we should even remove userConsentId, IMO we don't really need it.

guymoyo commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/CachedUserConsent.java:
```diff
@@ -40,6 +41,7 @@ public CachedUserConsent(UserConsentModel consentModel) {
         }
         this.createdDate = consentModel.getCreatedDate();
         this.lastUpdatedDate = consentModel.getLastUpdatedDate();
+        this.userConsentId = consentModel.getUserConsentId();
```
true, I think we should even remove userConsentId, IMO we don't really need it.

guymoyo commented:
Hi @tnorimat, I made some changes. I added the possibility to select 'grant_id_supported' in the admin UI, also I changed the logic of this grant flow, IMO I think it makes more sense to go with 'RequiredActionProvider', WDYT? 

There is some cross logic with the RAR, you will see that as comments in code.

tnorimat commented:
Hello @guymoyo ,

As you've proposed, it is good to use Required Action instead of coding  onto login service flow.

One point I would like to notice is as follows.

As discussed previously, to implement newly introduced User Grant, the following relationship is assumed.
```
Consent 1 --- 1 Grant
```

The current keycloak's Consent  has the following relationship.
```
{Client, User} 1 --- 1 Consent
```

Therefore, the following relationship is newly introduced.
```
{Client, User} 1 --- 1 Consent 1 --- 1 Grant
```

I'm afraid that what you want to do by Grant Management is achieved under this relationship.

guymoyo commented:
Hello @tnorimat,
Yes, you're right, consent is an agreement between a User and a Client. And the grant will extend that with fine-grained authorization. Therefore when the consent is canceled, all it's extended should be cancel/delete. 

I would like to try this:

grant representation
`     {
		"grantId" : ""
		"consentId" : "" // should be not null
		"clientId" : ""  // for quick access
		"userId" : ""   // for quick access
		"authorization_details" : {}
		"scopes" : []   // this scopes should be less than consent->grantedClientScopes
		"claims" : []   // same for the claims
		"createdDate" ""
		"lastUpdatedDate" ""
	},`

if we delete a consent, all grants with that consent will be deleted.

When creating the grant, if there is not yet consent between that Client and that User, it will be created implicitly.

tnorimat commented:
Hello @guymoyo ,

Thank you for your clarification.

guymoyo commented:
TODO:

- Incorporates the updates from the latest grant management publication (23.06.01)
- Avoid modifying RealmModel directly and follow CibaConfig for the spec param
- GrantEndpoint should follow --> CibaRootEndpoint concept
- Code redundancies for checking should be removed and make usage of [this pull request](https://github.com/Hitachi/keycloak/pull/2)
- Make grant and consent working like described in a comment upper 
- Write test
- write documentation 

Hi @tnorimat, here some tasks that I will continue working on, please feel free to add anything pertinent.

guymoyo commented:
Hi @tnorimat, I made some refactoring, can you please check again?

tnorimat commented:
Hello @guymoyo ,
Yes, I will check your revised codes.

tnorimat commented:
@guymoyo I will check them completely, but I have one suggestion at this time.

As the same as PAR and CIBA, it might be better for this grant management to be provided as preview feature.

WDYT?

guymoyo commented:
> @guymoyo I will check them completely, but I have one suggestion at this time.
> 
> As the same as PAR and CIBA, it might be better for this grant management to be provided as preview feature.
> 
> WDYT?

Yes, you're right, I forgot to change. I will change.

guymoyo commented on services/src/main/java/org/keycloak/authentication/requiredactions/GrantManagement.java:
```diff
@@ -0,0 +1,248 @@
+                userGrantModel.setUserId(user.getId());
+                userGrantModel.setScopes(grantScopes);
+                userGrantModel.setClaims(authSession.getAuthNote(OIDCLoginProtocol.CLAIMS_PARAM));
+                //TODO: when merging with RAR, userGrantModel.setAuthorizationDetails(rarProcessor.finaliseAuthorizationDetails(grantManagementAction, formData, authSession.getAuthNote(OIDCLoginProtocol.AUTHORIZATION_DETAILS_PARAM)));
```
@tnorimat How can we deal here, because grant management needs RAR to complete.

guymoyo commented on services/src/main/java/org/keycloak/authentication/requiredactions/GrantManagement.java:
```diff
@@ -0,0 +1,248 @@
+            try {
+                userGrantModel.setScopes(grantScopes);
+                userGrantModel.setClaims(authSession.getAuthNote(OIDCLoginProtocol.CLAIMS_PARAM));
+                //TODO: when merging with RAR, userGrantModel.setAuthorizationDetails(rarProcessor.finaliseAuthorizationDetails(grantManagementAction, formData, authSession.getAuthNote(OIDCLoginProtocol.AUTHORIZATION_DETAILS_PARAM)));
```
@tnorimat How can we deal here, because grant management needs RAR to complete.

tnorimat replied:
If supporting RAR at first is acceptable, we can resume grant management support after RAR is supported.
If not, could we skip this point? It means that grant management is partially supported at the first time.

tnorimat replied:
So, It seems that RAR is supported at first as you've working on https://github.com/tnorimat/keycloak/pull/24 .
Could you treat your RAR support at first before this grant management support?

guymoyo replied:
understand.

guymoyo commented on services/src/main/java/org/keycloak/protocol/oidc/grants/management/GrantManagementRootEndpoint.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return
+     */
+    @Path("query")
```
@tnorimat, the request to query a grant should be something like this 

`https://as.example.com/grants/TSdqirmAxDa0_-DB_1bASQ` and not `https://as.example.com/grants/query/TSdqirmAxDa0_-DB_1bASQ`

but following the CIBA endpoint concept, I have no choice to give a path here.
do you have an idea how can we solve that?

tnorimat replied:
I'm not sure but does the following work?
```
    @Path("/{grant_id}")
    @GET
```

guymoyo replied:
no, it doesn’t work.

tnorimat replied:
I see, I will find other solution.

guymoyo commented on services/src/main/java/org/keycloak/protocol/oidc/grants/management/GrantManagementRootEndpoint.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return
+     */
+    @Path("revoke")
```
same as `@Path("query")`

tnorimat replied:
I'm not sure but does the following work?
```
    @Path("/{grant_id}")
    @DELETE
```

guymoyo replied:
same, it doesn’t work.

tnorimat replied:
I see, I will find other solution.

tnorimat commented on server-spi-private/src/main/java/org/keycloak/models/utils/DefaultRequiredActions.java:
```diff
@@ -83,6 +83,17 @@ public static void addActions(RealmModel realm) {
             realm.addRequiredActionProvider(termsAndConditions);
         }
 
+        if (realm.getRequiredActionProviderByAlias(UserModel.RequiredAction.GRANT_MANAGEMENT.name()) == null) {
```
Considering that Grant Management might be preview feature, it might be good if it is not included as default and keycloak admin explicitly add corresponding required action afterwards. WDYT?

guymoyo replied:
true, I will remove.

tnorimat commented on server-spi-private/src/main/java/org/keycloak/protocol/LoginProtocol.java:
```diff
@@ -63,7 +63,12 @@
+        /**
+         * grant_id is unknown
+         */
+        INVALID_GRANT_ID,
```
I'm not sure where it is called. If it is called from only services projects, it might be better to move it to service project's some "org.keycloak.protocol.oidc.grants.management" package.

guymoyo replied:
It is not use, I forgot to remove.

tnorimat commented on server-spi/src/main/java/org/keycloak/models/RealmModel.java:
```diff
@@ -29,11 +29,9 @@
 import org.keycloak.storage.client.ClientStorageProviderModel;
 import org.keycloak.storage.role.RoleStorageProvider;
 import org.keycloak.storage.role.RoleStorageProviderModel;
```
By keycloak's convention, it might be better not to modify codes that is not related to the PR itself.

guymoyo replied:
Yes, I will change it.

tnorimat commented on server-spi/src/main/java/org/keycloak/models/UserModel.java:
```diff
@@ -300,7 +300,7 @@ default long getGroupsCountByNameContaining(String search) {
     enum RequiredAction {
         VERIFY_EMAIL, UPDATE_PROFILE, CONFIGURE_TOTP, UPDATE_PASSWORD, TERMS_AND_CONDITIONS,
-        VERIFY_PROFILE
+        VERIFY_PROFILE, GRANT_MANAGEMENT
```
Considering that Grant Management might be preview feature at first, it might be better not to include it inside `UserModel` like `WebAuthnRegister` required action.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/grants/management/DefaultGrantManagementProvider.java:
```diff
@@ -0,0 +1,153 @@
+package org.keycloak.protocol.oidc.grants.management;
+
+
+import org.keycloak.models.*;
```
By keycloak's convention, it might be better not to use "*" import directive.

guymoyo replied:
Yes, I will change it.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/grants/management/GrantManagementQueryEndpoint.java:
```diff
@@ -0,0 +1,76 @@
+import org.keycloak.common.Profile;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.*;
```
By keycloak's convention, it might be better not to use "*" import directive.

guymoyo replied:
Yes, I will change it.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/grants/management/GrantManagementQueryEndpoint.java:
```diff
@@ -0,0 +1,76 @@
+import org.keycloak.services.resources.Cors;
+import org.keycloak.utils.ProfileHelper;
+
+import javax.ws.rs.*;
```
By keycloak's convention, it might be better not to use "*" import directive.

guymoyo replied:
Yes, I will change it

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/grants/management/GrantManagementRevokeEndpoint.java:
```diff
@@ -0,0 +1,96 @@
+import org.keycloak.common.Profile;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.*;
```
By keycloak's convention, it might be better not to use "*" import directive.

guymoyo replied:
Yes, I will change it.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/grants/management/GrantManagementRevokeEndpoint.java:
```diff
@@ -0,0 +1,96 @@
+import org.keycloak.services.resources.Cors;
+import org.keycloak.utils.ProfileHelper;
+
+import javax.ws.rs.*;
```
By keycloak's convention, it might be better not to use "*" import directive.

guymoyo replied:
Yes, I will change it.

tnorimat reviewed (commented):
@guymoyo I've added some review comments. Could you check them?

guymoyo commented:
Yes, I will change it.

guymoyo commented on server-spi-private/src/main/java/org/keycloak/models/GrantManagementProvider.java:
```diff
@@ -0,0 +1,35 @@
+
+    UserGrantModel getGrantByGrantId(RealmModel realm, String grantId) throws Exception;
+
+    UserGrantModel getGrantByGrantIdAndClientId(RealmModel realm, String grantId, String clientId) throws Exception;
```
getGrantByGrantId is not use anymore, I will remove.

guymoyo commented on server-spi-private/src/main/java/org/keycloak/models/Constants.java:
```diff
@@ -127,4 +128,17 @@
+    public static final String GRANT_MANAGEMENT_ACTION_UPDATE = "update";
+    public static final String GRANT_MANAGEMENT_ACTION_REPLACE = "replace";
+    public static final String GRANT_MANAGEMENT_ACTION_CREATE = "create";
+    public static final List<String> GRANT_MANAGEMENT_ACTIONS = Arrays.asList(GRANT_MANAGEMENT_ACTION_QUERY, GRANT_MANAGEMENT_ACTION_REVOKE, GRANT_MANAGEMENT_ACTION_UPDATE, GRANT_MANAGEMENT_ACTION_REPLACE, GRANT_MANAGEMENT_ACTION_CREATE);
```
I will remove that here and create a class inside "org.keycloak.protocol.oidc.grants.management" package

guymoyo commented:
Hi @tnorimat, I made the changes but still having this problem with the path of the grant management endpoint.

tnorimat commented:
@guymoyo As I added comments, I will find other solution for the final issue.

guymoyo commented:
Hi @tnorimat, may this help to have a good overview on [Relation Grant vs Consent](https://bitbucket.org/openid/fapi/issues/442/grant-vs-consent-confusion)

tnorimat commented:
Hello @guymoyo ,

Thank you for your reference https://bitbucket.org/openid/fapi/issues/442/grant-vs-consent-confusion ,
I will follow this discussion thread. I've looked it at a glance and found that it might be needed to discuss the definition of Consent and Grant by considering their nature and  the relationship between them. 

From your implementer's side, is it better to have relationship Consent 1 --- * Grant ?

guymoyo commented:
> Hello @guymoyo ,
> 
> Thank you for your reference https://bitbucket.org/openid/fapi/issues/442/grant-vs-consent-confusion ,
> I will follow this discussion thread. I've looked it at a glance and found that it might be needed to discuss the definition of Consent and Grant by considering their nature and the relationship between them.
> 
> From your implementer's side, is it better to have relationship Consent 1 --- * Grant ?

Yes from the implementer's side, I think it is better to go in this way and it is actually implemented like that in this branch. but it can be changed if we have another clear relation view.
